src/SonarScanner.MSBuild.Tasks/Targets/SonarQube.Integration.targets

```diff
@@ -149,26 +150,44 @@
     <Error Text="SonarQube analysis is only supported with MSBuild 14 or later." />
   </Target>
 
+  <Target Name="SonarResolveReferencesMsBuild14"
+          AfterTargets="ResolveAssemblyReferences"
+          Condition=" $(SonarQubeTempPath) != '' AND $(MSBuildToolsVersion) == '14.0'">
+    <ItemGroup>
+      <SonarResolvedReferences Include="@(%(ReferencePath.FusionName))" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="SonarResolveReferences"
+          AfterTargets="FindReferenceAssembliesForReferences"
+          Condition=" $(SonarQubeTempPath) != '' AND $(MSBuildToolsVersion) != '14.0'">
+    <ItemGroup>
+      <SonarResolvedReferences Include="@(%(ReferencePathWithRefAssemblies.FusionName))" />
+    </ItemGroup>
+  </Target>
+
   <!-- **************************************************************************** -->
   <!-- Categorize as a product or test project -->
   <!-- **************************************************************************** -->
-  <!-- Work out if the project is a test or Fakes project. -->
-  <!-- Fakes projects are auto-generated. If a project is recognized as a Fakes
-       project then it will be marked as a test project and excluded.
-       For test projects, if it is explicitly marked then use that value, otherwise:
-          * check if the name matches the regular expression in the config file, or
-          * check if it has one of the expected project types guids.
-         If either is true then the project is a test project, otherwise
-         it is a product project.
-
-         This target is a utility target: it won't be executed unless a target that
-         depends on it is executed.
+  <!-- Project is excluded if it's recognized as a Fakes or a Temp project. These projects are auto-generated.
+       Project is categorized as a Test project if:
+        * Is explicitly marked with SonarQubeTestProject=true
+        * Is a Fakes or a Temp project explicitly marked as SonarQubeExclude=false
+        * Has one of the expected test project GUIDs.
+        * Has a TestContainer project capability.
+        * References a known unit test related assembly.
```

Comment:
I know that some projects (not unit tests) can reference test assemblies : how do you deal with them ?

Reply:
In old state:
* S4MSB classifies them as PRODUCT
* Analyzer classifies them as TEST and run only rules with scope TEST.
* Users have no way out of this to get MAIN rules.

In current state (MMF-2298):
* S4MSB classifies them as test, no analysis is done
* User can opt-out using `<SonarQubeTestProject>false<SonarQubeTestProject>`, analyzer will respect that

In a 5.2 release (after MMF-2297 starting next week)
* S4MSB classifies them as test **and runs the analysis**
* Analyzer respects the S4MSB decision and runs rules with scope TEST.
* User can override the scope using `<SonarQubeTestProject>false<SonarQubeTestProject>` to get MAIN rules.